services/worker-manager/src/data.js

```diff
@@ -134,56 +133,157 @@ class WorkerPool {
-  rowKey: Entity.keys.StringKey('errorId'),
-  properties: {
-    // The worker pool this maps to.
-    workerPoolId: Entity.types.String,
-
-    // An arbitrary id for this error
-    errorId: Entity.types.SlugId,
-
-    // The datetime this error occured
-    reported: Entity.types.Date,
-
-    // The sort of error this is. Can be used by UIs to differentiate
-    kind: Entity.types.String,
-
-    // A human readable name for this error
-    title: Entity.types.String,
-
-    // A human readable description of this error and what can be done to fix it
-    description: Entity.types.String,
-
-    // Anything else that a reporter may want to add in a structured way
-    extra: Entity.types.JSON,
-  },
-});
-
-WorkerPoolError.prototype.serializable = function() {
-  return {
-    workerPoolId: this.workerPoolId,
-    errorId: this.errorId,
-    reported: this.reported.toJSON(),
-    kind: this.kind,
-    title: this.title,
-    description: this.description,
-    extra: this.extra,
-  };
-};
+class WorkerPoolError {
+  // (private constructor)
+  constructor(props) {
+    Object.assign(this, props);
+  }
 
-WorkerPoolError.expire = async function(threshold) {
-  await this.scan({
-    reported: Entity.op.lessThan(threshold),
-  }, {
-    limit: 500,
-    handler: async item => {
-      await item.remove();
-    },
-  });
-};
+  // Create a single instance from a DB row
+  static fromDb(row) {
+    return new WorkerPoolError({
+      errorId: row.error_id,
+      workerPoolId: row.worker_pool_id,
+      reported: row.reported,
+      kind: row.kind,
+      title: row.title,
+      description: row.description,
+      extra: row.extra,
+    });
+  }
+
+  // Create a single instance, or undefined, from a set of rows containing zero
+  // or one elements.  This matches the semantics of get_worker_pool_error.
+  static fromDbRows(rows) {
+    if (rows.length === 1) {
+      return WorkerPool.fromDb(rows[0]);
+    }
+  }
+
+  // Create an instance from API arguments, with default values applied.
+  static fromApi(input) {
+    const now = new Date();
+    return new WorkerPoolError({
+      extra: {},
+      reported: now,
+      ...input,
+    });
+  }
+
+  // Get a worker pool error from the DB, or undefined if it does not exist.
+  static async get(db, errorId) {
+    return WorkerPoolError.fromDbRows(await db.fns.get_worker_pool_error(errorId));
+  }
+
+  // Call db.get_worker_pool_errors.
+  // You can use this in two ways: with a handler or without a handler.
+  // In the latter case you'll get a list of up to 1000 entries and a
+  // continuation token.
+  // The response will be of the form { rows, continationToken }.
+  // If there are no workers to show, the response will have the
+  // `rows` field set to an empty array.
+  //
+  // If a handler is supplied, then it will invoke the handler
+  // on every item of the scan.
+  static async getWorkerPoolErrors(
+    db,
+    {
+      query,
+      handler,
```

Comment:
The handler is not used so no need to have this.

Reply:
Removed the handler portion.